java/org/apache/catalina/TomcatPrincipal.java

```diff
@@ -47,4 +48,58 @@
      *                   exception to LoginContext
      */
     void logout() throws Exception;
+
+    /**
+     * Returns the value of the named attribute as an <code>Object</code>, or
+     * <code>null</code> if no attribute of the given name exists. May also
+     * return <code>null</code>, if the named attribute exists but cannot be
+     * returned in a way that ensures that changes made to the returned object
+     * are not reflected by objects returned by subsequent calls to this method.
+     * <p>
+     * Only the servlet container may set attributes to make available custom
+     * information about a Principal or the user it represents. For example,
+     * some of the Realm implementations can be configured to additionally query
+     * user attributes from the <i>user database</i>, which then are provided
+     * through the Principal's attributes map.
+     * <p>
+     * In order to keep the attributes map <i>immutable</i>, the objects
+     * returned by this method should always be <i>defensive copies</i> of the
+     * objects contained in the attributes map. Any changes applied to these
+     * objects must not be reflected by objects returned by subsequent calls to
+     * this method. If that cannot be guaranteed (e. g. there is no way to copy
+     * the object), the object's string representation (or even
+     * <code>null</code>) shall be returned.
+     * <p>
+     * Attribute names and naming conventions are maintained by the Tomcat
+     * components that contribute to this map, like some of the Realm
+     * implementations.
+     *
+     * @param name a <code>String</code> specifying the name of the attribute
```

Comment:
You should specify that the name must not be null and will throw NPE.

Reply:
Ah, ok. Basically, I was just copying from `jakarta.servlet.ServletRequest#getAttribute(String)`. You prefer a real `@exception` tag? None of these many `getAttribute(String)` used in Servlets mention that explicitly.

Reply:
I would prefer explicit documentation of behavior rather than trial and error. Since the servlet spec does not tell, maybe `null` is return for `null`. Maybe @markt-asf can tell. I would make the consistent even if the servlet spec approach feels wrong.

Reply:
Now, I understand... It only NPEs, if `CaseInsensitiveKeyMap` is used, which does not support `null` as a key (and that's documented). I will add an explicit check
```
if (name == null)  {
    return null;
}
```
to method `getAttribute(String)` and mention that in this method's Javadoc comment accordingly.

Reply:
That's fine.